### PR TITLE
[PDI-19580]-Enable Use of OAUTH for PDI Mail Steps and Entries that use IMAP/SMTP

### PR DESCRIPTION
**Pull request by Mandakini111**
[PDI-19580]-Enable Use of OAUTH for PDI Mail Steps and Entries that use IMAP/SMTP

[PDI-19580]: https://hv-eng.atlassian.net/browse/PDI-19580?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by buildguy** [comment]
<div align='center'>

[🚨 Frogbot scanned this pull request and found the below:](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>




---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
[comment]: <> (FrogbotReviewComment)

```
db.parse( inputSource )
```
at `core/src/main/java/org/pentaho/di/core/xml/XMLHandler.java` (line 749)


[comment]: <> (FrogbotReviewComment)


## 🎯 Static Application Security Testing (SAST) Vulnerability
<div align='center'>

| Severity                | Finding                  |
| :---------------------: | :-----------------------------------: |
| <br>  Medium | XML external entities |

</div>

<details>
<summary> <b>Full description</b> </summary>
<br>


### Overview
Unsafe XXE (XML External Entity) expansion, is a type of vulnerability
that allows an attacker to access external resources, such as files or
network services using a crafted XML document. The attack can also lead to
a denial of service, due to quadratic expansion of XXE elements.

### Vulnerable example
```java
protected void doPost(HttpServletRequest request, HttpServletResponse response)
    throws ServletException, IOException {
    DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
    DocumentBuilder db = dbf.newDocumentBuilder();
    InputSource is = new InputSource();
    // Load the xml string into an InputSource object.
    is.setCharacterStream(new StringReader( request.getParameter("xml") ));
    // Parse it
    db.parse(is);
}
```
### Remediation
Set the following features on `DocumentBuilderFactory` to remediate XXE:
```diff
+   try {
+      dbf.setFeature("http://xml.org/sax/features/external-general-entities", false);
+   } catch (ParserConfigurationException e) { }
+   try {
+      dbf.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+   } catch (ParserConfigurationException e) { }
+   try {
+      dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+   } catch (ParserConfigurationException e) { }
+   try {
+      dbf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd",
+           false);
+   } catch (ParserConfigurationException e) { }
+   dbf.setXIncludeAware(false);
+   dbf.setExpandEntityReferences(false);
    DocumentBuilder db = dbf.newDocumentBuilder();
```


</details>

<details>
<summary> <b>Code Flows</b> </summary>
<br>


<details>
<summary> <b>Vulnerable data flow analysis result</b> </summary>
<br>


↘️ `inputStream.read()` (at engine/src/main/java/org/pentaho/di/pan/PanCommandExecutor.java line 380)

↘️ `c` (at engine/src/main/java/org/pentaho/di/pan/PanCommandExecutor.java line 380)

↘️ `xml.append( (char) c )` (at engine/src/main/java/org/pentaho/di/pan/PanCommandExecutor.java line 381)

↘️ `xml.toString()` (at engine/src/main/java/org/pentaho/di/pan/PanCommandExecutor.java line 384)

↘️ `XMLHandler.loadXMLString( xml.toString() )` (at engine/src/main/java/org/pentaho/di/pan/PanCommandExecutor.java line 384)

↘️ `DocumentBuilder db` (at core/src/main/java/org/pentaho/di/core/xml/XMLHandler.java line 741)

↘️ `db.parse( inputSource )` (at core/src/main/java/org/pentaho/di/core/xml/XMLHandler.java line 749)


</details>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :x: Build failed in 2h 10m 11s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
plugins/email-messages/impl,plugins/email-messages/ui,plugins/mail-job/impl,plugins/mail-job/ui
```
---
:ok_hand: All tests passed!
 
Tests run: **47**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9612/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by lucboudreau** [comment]
Is it normal that nothing is encrypted here? I would expect all secret keys to be encrypted with the pdi encryption module.

**Review by lucboudreau — CHANGES_REQUESTED** [review]
All secrets should use the pdi encryption module

**Comment by buildguy** [comment]
<div align='center'>

[🚨 Frogbot scanned this pull request and found the below:](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>




---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
[comment]: <> (FrogbotReviewComment)

```
new File( filename )
```
at `core/src/main/java/org/pentaho/di/laf/OverlayProperties.java` (line 60)


[comment]: <> (FrogbotReviewComment)


## 🎯 Static Application Security Testing (SAST) Vulnerability
<div align='center'>

| Severity                | Finding                  |
| :---------------------: | :-----------------------------------: |
| <br>  Medium | Stored Path Traversal |

</div>

<details>
<summary> <b>Full description</b> </summary>
<br>


### Overview
Path traversal, also known as directory traversal, is a type of
vulnerability that allows an attacker to access files or directories on a
computer or device that are outside of the intended directory.
Allowing arbitrary read access can allow the attacker to read sensitive
files, such as configuration files or sensitive data, potentially leading
data loss or even system compromise. Allowing arbitrary write access is
more severe and in most cases leads to arbitrary code execution, via
editing important system files or sensitive data.

### Vulnerable example
```java
public class path_traversaLvuln {
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String DOCS_FOLDER = "/srv/www/docs";
        String docName = statement.executeQuery(query);  // Reading from DB
        Path docPath = Paths.get(DOCS_FOLDER, docName);
        File docFile = docPath.toFile();
        FileUtils.copyFile(docFile, response.getOutputStream());
    }
}
```
In this example, an attacker can, via a stored parameter, inject a back-path,
that will get anywhere in the system, using "../../".

### Remediation
```diff
public class path_traversal_safe {
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String DOCS_FOLDER = "/srv/www/docs";
        String docName = statement.executeQuery(query);  // Reading from DB
        Path docPath = Paths.get(DOCS_FOLDER, docName);
        + Path normDocPath = docPath.normalize();
        + // Make sure the canonical path resides in the desired dir
        + if (normDocPath.startsWith(DOCS_FOLDER)) {
            File docFile = docPath.toFile();
            FileUtils.copyFile(docFile, response.getOutputStream());
        + }
    }
}
```
By checking that the folder name still starts with the predefined prefix, we
make sure that the attacker is not able to back-path outside of the allowed
folder.


</details>

<details>
<summary> <b>Code Flows</b> </summary>
<br>


<details>
<summary> <b>Vulnerable data flow analysis result</b> </summary>
<br>


↘️ `System.getProperty( variable, deflt )` (at core/src/main/java/org/pentaho/di/core/Const.java line 2641)

↘️ `return System.getProperty( variable, deflt );` (at core/src/main/java/org/pentaho/di/core/Const.java line 2641)

↘️ `Const.getEnvironmentVariable( "org.pentaho.di.laf.alt", null )` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 54)

↘️ `altFile` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 54)

↘️ `loadProps( altFile )` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 56)

↘️ `String filename` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 86)

↘️ `new OverlayProperties( filename )` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 88)

↘️ `public class OverlayProperties extends Properties implements PropertyHandler {

  private static final long serialVersionUID = 1L;
  private String name = null;

  public OverlayProperties( String file ) throws IOException {
    load( file );
  }

  @Override
  public boolean exists( String filename ) {
    try {
      return ( getURL( filename ) != null );
    } catch ( MalformedURLException e ) {
      return false;
    }
  }

  @Override
  public boolean loadProps( String filename ) {
    try {
      return load( filename );
    } catch ( IOException e ) {
      return false;
    }
  }

  private URL getURL( String filename ) throws MalformedURLException {
    URL url;
    File file = new File( filename );
    if ( file.exists() ) {
      url = file.toURI().toURL();
    } else {
      ClassLoader classLoader = getClass().getClassLoader();
      url = classLoader.getResource( filename );
    }
    return url;
  }

  /**
   * cleanse and reload the property file
   *
   * @param filename
   * @return
   * @throws IOException
   */
  public boolean load( String filename ) throws IOException {
    URL url = getURL( filename );
    if ( url == null ) {
      return false;
    }
    clear();
    load( url.openStream() );
    return true;
  }

  public String getName() {
    return name;
  }
}` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 31)

↘️ `String file` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 36)

↘️ `load( file )` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 37)

↘️ `String filename` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 77)

↘️ `getURL( filename )` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 78)

↘️ `String filename` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 58)

↘️ `new File( filename )` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 60)


</details>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
[comment]: <> (FrogbotReviewComment)

```
classLoader.getResource( filename )
```
at `core/src/main/java/org/pentaho/di/laf/OverlayProperties.java` (line 65)


[comment]: <> (FrogbotReviewComment)


## 🎯 Static Application Security Testing (SAST) Vulnerability
<div align='center'>

| Severity                | Finding                  |
| :---------------------: | :-----------------------------------: |
| <br>  Medium | Stored Path Traversal |

</div>

<details>
<summary> <b>Full description</b> </summary>
<br>


### Overview
Path traversal, also known as directory traversal, is a type of
vulnerability that allows an attacker to access files or directories on a
computer or device that are outside of the intended directory.
Allowing arbitrary read access can allow the attacker to read sensitive
files, such as configuration files or sensitive data, potentially leading
data loss or even system compromise. Allowing arbitrary write access is
more severe and in most cases leads to arbitrary code execution, via
editing important system files or sensitive data.

### Vulnerable example
```java
public class path_traversaLvuln {
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String DOCS_FOLDER = "/srv/www/docs";
        String docName = statement.executeQuery(query);  // Reading from DB
        Path docPath = Paths.get(DOCS_FOLDER, docName);
        File docFile = docPath.toFile();
        FileUtils.copyFile(docFile, response.getOutputStream());
    }
}
```
In this example, an attacker can, via a stored parameter, inject a back-path,
that will get anywhere in the system, using "../../".

### Remediation
```diff
public class path_traversal_safe {
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String DOCS_FOLDER = "/srv/www/docs";
        String docName = statement.executeQuery(query);  // Reading from DB
        Path docPath = Paths.get(DOCS_FOLDER, docName);
        + Path normDocPath = docPath.normalize();
        + // Make sure the canonical path resides in the desired dir
        + if (normDocPath.startsWith(DOCS_FOLDER)) {
            File docFile = docPath.toFile();
            FileUtils.copyFile(docFile, response.getOutputStream());
        + }
    }
}
```
By checking that the folder name still starts with the predefined prefix, we
make sure that the attacker is not able to back-path outside of the allowed
folder.


</details>

<details>
<summary> <b>Code Flows</b> </summary>
<br>


<details>
<summary> <b>Vulnerable data flow analysis result</b> </summary>
<br>


↘️ `System.getProperty( variable, deflt )` (at core/src/main/java/org/pentaho/di/core/Const.java line 2641)

↘️ `return System.getProperty( variable, deflt );` (at core/src/main/java/org/pentaho/di/core/Const.java line 2641)

↘️ `Const.getEnvironmentVariable( "org.pentaho.di.laf.alt", null )` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 54)

↘️ `altFile` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 54)

↘️ `loadProps( altFile )` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 56)

↘️ `String filename` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 86)

↘️ `new OverlayProperties( filename )` (at core/src/main/java/org/pentaho/di/laf/OverlayPropertyHandler.java line 88)

↘️ `public class OverlayProperties extends Properties implements PropertyHandler {

  private static final long serialVersionUID = 1L;
  private String name = null;

  public OverlayProperties( String file ) throws IOException {
    load( file );
  }

  @Override
  public boolean exists( String filename ) {
    try {
      return ( getURL( filename ) != null );
    } catch ( MalformedURLException e ) {
      return false;
    }
  }

  @Override
  public boolean loadProps( String filename ) {
    try {
      return load( filename );
    } catch ( IOException e ) {
      return false;
    }
  }

  private URL getURL( String filename ) throws MalformedURLException {
    URL url;
    File file = new File( filename );
    if ( file.exists() ) {
      url = file.toURI().toURL();
    } else {
      ClassLoader classLoader = getClass().getClassLoader();
      url = classLoader.getResource( filename );
    }
    return url;
  }

  /**
   * cleanse and reload the property file
   *
   * @param filename
   * @return
   * @throws IOException
   */
  public boolean load( String filename ) throws IOException {
    URL url = getURL( filename );
    if ( url == null ) {
      return false;
    }
    clear();
    load( url.openStream() );
    return true;
  }

  public String getName() {
    return name;
  }
}` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 31)

↘️ `String file` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 36)

↘️ `load( file )` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 37)

↘️ `String filename` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 77)

↘️ `getURL( filename )` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 78)

↘️ `String filename` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 58)

↘️ `classLoader.getResource( filename )` (at core/src/main/java/org/pentaho/di/laf/OverlayProperties.java line 65)


</details>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>



---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 3h 19m 6s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
plugins/email-messages/impl,plugins/email-messages/ui,plugins/mail-job/impl,plugins/mail-job/ui
```
---
:ok_hand: All tests passed!
 
Tests run: **47**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9612/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>



---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 1h 38m 45s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
plugins/email-messages/impl,plugins/email-messages/ui,plugins/mail-job/impl,plugins/mail-job/ui
```
---
:ok_hand: All tests passed!
 
Tests run: **47**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9612/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: plugins/email-messages/impl/pom.xml [hunk 1]
```diff
@@ -49,5 +49,11 @@
       <classifier>tests</classifier>
       <scope>test</scope>
     </dependency>
+      <dependency>
```
Comment (peterrinehart) [inline plugins/email-messages/impl/pom.xml:52]: This won't work in our build pipeline.  pentaho-platform extensions gets built here: https://github.com/pentaho/jenkins-pipelines/blob/10.3.0.0/resources/builders/suite-release.yaml#L329

This plugin gets built as part of the kettle lowdeps here, which is earlier: https://github.com/pentaho/jenkins-pipelines/blob/10.3.0.0/resources/builders/suite-release.yaml#L209

Either this dependency needs to be removed or this plugin needs to be moved to the highdeps section.

File: plugins/email-messages/impl/src/test/java/org/pentaho/di/job/entries/getpop/JobEntryGetPOPTest.java [hunk 2]
```diff
@@ -21,10 +21,10 @@
...
 package org.pentaho.di.job.entries.getpop;
 
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
-
+import static org.junit.Assert.*;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
```
Comment (peterrinehart) [inline plugins/email-messages/impl/src/test/java/org/pentaho/di/job/entries/getpop/JobEntryGetPOPTest.java:27]: Please avoid importing with .* and only import the needed classes.

File: plugins/email-messages/impl/src/test/java/org/pentaho/di/job/entries/getpop/JobEntryGetPOPTest.java [hunk 3]
```diff
@@ -320,4 +337,37 @@ public void testEnvVariablesAreSubstitutedForFolders() {
...
+    assertEquals( JobEntryGetPOP.GRANTTYPE_CLIENTCREDENTIALS, entry.getGrant_type() );
+  }
+
+  @Test
+  public void testgrantTypeIsAuthorizationCode() {
+    entry.setGrant_type( JobEntryGetPOP.GRANTTYPE_AUTHORIZATION_CODE );
+    assertEquals( JobEntryGetPOP.GRANTTYPE_AUTHORIZATION_CODE, entry.getGrant_type() );
+  }
+
+  @Test( expected = RuntimeException.class )
```
Comment (peterrinehart) [inline plugins/email-messages/impl/src/test/java/org/pentaho/di/job/entries/getpop/JobEntryGetPOPTest.java:352]: Can we test for something more selective than a RuntimeException?  A lot of things could possibly cause this test to look like it works when it did not.

File: plugins/email-messages/impl/src/test/java/org/pentaho/di/job/entries/getpop/JobEntryGetPOPTest.java [hunk 4]
```diff
@@ -320,4 +337,37 @@ public void testEnvVariablesAreSubstitutedForFolders() {
...
+    StatusLine statusLine = new BasicStatusLine( new ProtocolVersion( "HTTP", 1, 1 ), 400, "Bad Request" );
+
+    when( httpClientManager.createDefaultClient() ).thenReturn( httpClient );
+    when( httpClient.execute( any( HttpPost.class ) ) ).thenReturn( httpResponse );
+    when( httpResponse.getStatusLine() ).thenReturn( statusLine );
+
+    entry.getOauthToken( tokenUrl );
+  }
+
+  @Test( expected = RuntimeException.class )
```
Comment (peterrinehart) [inline plugins/email-messages/impl/src/test/java/org/pentaho/di/job/entries/getpop/JobEntryGetPOPTest.java:365]: Same comment as above.

File: plugins/mail-job/impl/pom.xml [hunk 5]
```diff
@@ -55,5 +55,11 @@
       <classifier>tests</classifier>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>pentaho</groupId>
+      <artifactId>pentaho-platform-extensions</artifactId>
```
Comment (peterrinehart) [inline plugins/mail-job/impl/pom.xml:60]: Same comment as the first one; this dependency cannot be here unless this plugin moves to the highdeps section.

File: plugins/mail-job/impl/src/test/java/org/pentaho/di/job/entries/mail/JobEntryTest.java [hunk 6]
```diff
@@ -63,4 +97,44 @@ public void testJobEntrymailPasswordEncrVar() {
...
+    assertEquals( JobEntryMail.GRANTTYPE_CLIENTCREDENTIALS, jobEntryMail.getGrant_type() );
+  }
+
+  @Test
+  public void testgrantTypeIsAuthorizationCode() {
+    jobEntryMail.setGrant_type( JobEntryMail.GRANTTYPE_AUTHORIZATION_CODE );
+    assertEquals( JobEntryMail.GRANTTYPE_AUTHORIZATION_CODE, jobEntryMail.getGrant_type() );
+  }
+
+  @Test( expected = RuntimeException.class )
```
Comment (peterrinehart) [inline plugins/mail-job/impl/src/test/java/org/pentaho/di/job/entries/mail/JobEntryTest.java:123]: Same as the comments in the other test; can we be more selective here?  Is there any other check we can do to ensure the test failed in the expected way?

**Review by peterrinehart — CHANGES_REQUESTED** [review]

**Pull request closed without merging**